adafruit_ble/scanner.py

```diff
@@ -0,0 +1,188 @@
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+"""
+`adafruit_ble.scanner`
+====================================================
+
+Scan for nearby BLE Devices
+
+* Author(s): Dan Halbert for Adafruit Industries
+
+"""
+import struct
+
+import bleio
+
+from .advertising import AdvertisingPacket
+
+class Scanner:
+    """
+    Scan for BLE device advertisements for a period of time. Return what was received.
+
+    Example::
+
+        from adafruit_ble.scanner import Scanner
+        scanner = Scanner()
+        scanner.Scan
+
+        # Wait for a connection.
+        while not uart.connected:
+            pass
+
+        uart.write('abc')
+    """
+
+    def __init__(self):
+        self._scanner = bleio.Scanner()
+
+    def scan_unique(self, timeout, *, interval=0.1, window=0.1):
+        """Scan for advertisements from BLE devices. Suppress duplicates
+        in returned `ScanEntry` objects.
+
+        :param int timeout: how long to scan for (in seconds)
+        :param float interval: the interval (in seconds) between the start
+           of two consecutive scan windows.
+           Must be in the range 0.0025 - 40.959375 seconds.
+        :param float window: the duration (in seconds) to scan a single BLE channel.
+          `window` must be <= `interval`.
+        :returns a list of `adafruit_ble.ScanEntry` objects.
+
+        """
+        return ScanEntry.unique(self.scan(timeout, interval=interval, window=window))
+
+    def scan(self, timeout, *, interval=0.1, window=0.1):
+        """Scan for advertisements from BLE devices.
+
+        :param int timeout: how long to scan for (in seconds)
+        :param float interval: the interval (in seconds) between the start
+           of two consecutive scan windows.
+           Must be in the range 0.0025 - 40.959375 seconds.
+        :param float window: the duration (in seconds) to scan a single BLE channel.
+          `window` must be <= `interval`.
+        :returns a list of `adafruit_ble.ScanEntry` objects.
+
+        """
+        return [ScanEntry(e) for e in self._scanner.scan(timeout, interval=interval, window=window)]
+
+class ScanEntry:
+    """
+    Information about an advertising packet from a BLE device received by a `Scanner`.
+
+    :param bleio.ScanEntry scan_entry: lower-level ScanEntry returned from `bleio.Scanner`.
+      This constructor is normally used only by `Scanner`.
+    """
+
+    def __init__(self, scan_entry):
+        self._bleio_scan_entry = scan_entry
+
+    def item(self, item_type):
```

Comment:
Should this be `__getitem__` so it works like a dictionary?

Reply:
Sounds more Pythonic to me. Will do.

Reply:
Actually, not sure about this. `__getitem__` should raise `KeyError` if not found. I just have it return `None`, which makes for simpler checking (see for instance `service_uuids`, which would have a lot of `try`s. What do you think?

Reply:
KeyError is nothing new to Python. The way around it is by 1) using ` in ` before getting a key or 2) using `.get` to provide a default. (https://docs.python.org/3/library/stdtypes.html?highlight=dictionary#dict.get). The latter you would have to provide but I think ` in ` should just work.

Reply:
RIght, I thought of using `get()` after I wrote the above. So I'll change it to be `__getitem__`.